syntax = "proto3";

package todo_app.v1;

import "resemble/v1alpha1/options.proto";

////////////////////////////////////////////////////////////////////////

service TodoList {
  option (resemble.v1alpha1.service).state = "TodoListState";

  rpc AddTodo(AddTodoRequest) returns (AddTodoResponse) {
    option (resemble.v1alpha1.method).writer = {
    };
  }

  rpc ListTodos(ListTodosRequest) returns (ListTodosResponse) {
    option (resemble.v1alpha1.method).reader = {
    };
  }

  rpc DeleteTodo(DeleteTodoRequest) returns (DeleteTodoResponse) {
    option (resemble.v1alpha1.method).writer = {
    };
  }

  rpc CompleteTodo(CompleteTodoRequest) returns (CompleteTodoResponse) {
    option (resemble.v1alpha1.method).writer = {
    };
  }
}

message Todo {
  string id = 1;
  string text = 2;
  bool complete = 3;
}

message TodoListState {
  repeated Todo todos = 1;
}

message AddTodoRequest {
  string todolistId = 1;
  string todo = 2;
}

message AddTodoResponse {}

message ListTodosRequest {
  string todolistId = 1;
}

message ListTodosResponse {
  repeated Todo todos = 1;
}

message DeleteTodoRequest {
  string todolistId = 1;
  string todoId = 2;
}

message DeleteTodoResponse {}

message CompleteTodoRequest {
  string todolistId = 1;
  string todoId = 2;
}

message CompleteTodoResponse {}